### ISSUE TITLE
Am I using the wrong function to project satellites?

### ISSUE THREAD

**Issue by owntheweb** [brandon-rhodes/python-sgp4#7]
Hello,
Pardon my lack of proper terminology as I'm more of an artist than a satellite guru. As a result, I'll visualize as much as I can below.

![sos](https://cloud.githubusercontent.com/assets/611860/5706870/92b31f1a-9a41-11e4-8b48-9091cbf0a3b8.jpg)

I'm having some starting success plotting 15,147 orbiting objects around Earth, projected for a volunteer Science On a Sphere project at the museum setting at my office.

![preview1](https://cloud.githubusercontent.com/assets/611860/5706357/55608eee-9a3d-11e4-84b6-aab8e7dfd864.png)

Cool!

The next challenge: Working on some alignment, comparing with ISS tle data, the orbit seems too "circular"/"ring-like", overlapping the same orbit over and over like this:

![preview2-orbit-issue](https://cloud.githubusercontent.com/assets/611860/5706389/cc0f102e-9a3d-11e4-88b4-76ab305187d0.png)

Here's what it looks like when I skip the re-projections and show in 3D space based on the values sgp4 is producing (a nice ring):

![screen shot 2015-01-12 at 9 44 13 am](https://cloud.githubusercontent.com/assets/611860/5706623/a09257b0-9a3f-11e4-9e6a-ffcbc9d683ee.png)

I'm used to seeing it look something like this (from a JavaScript project using [this library](https://github.com/shashwatak/satellite-js)):

![screen shot 2015-01-12 at 9 35 30 am](https://cloud.githubusercontent.com/assets/611860/5706478/81204b90-9a3e-11e4-9625-b2d8129bd6a4.png)

Is it possible that I'm using the wrong values or functions to project this?

Here's a simplified version of the script with how I'm producing the position values:

    from sgp4.earth_gravity import wgs72
    from sgp4.io import twoline2rv
    import datetime

    def sat_pos_vel(sat, y, m, d, h, i, s):
        position, velocity = sat.propagate(y, m, d, h, i, s)
        return {"p":position, "v":velocity}

    line1 = "1 25544U 98067A   15011.54761787  .00015493  00000-0  24903-3 0  7129"
    line2 = "2 25544 051.6471 145.9316 0006026 238.8534 213.9008 15.53184119923669"
    sat = twoline2rv(line1, line2, wgs72)

    when = datetime.datetime.now()

    for i in range(300):
        sat_prop = sat_pos_vel(sat, when.year, when.month, when.day, when.hour, when.minute, when.second)
        when = when + datetime.timedelta(0,100) #add 100 seconds

        print(sat_prop)

I'm using Python 3.4

Let me know if I can provide additional information.

Thanks for taking a look in advance! I truly appreciate it.

**Comment by owntheweb** [comment]
The answer may be here:
http://stackoverflow.com/questions/15362557/python-sgp4-1-1-calculating-incorrect-orbit

**Comment by brandon-rhodes** [comment]
You are entirely correct, @owntheweb — the issue you ran across explains the problem.

Per Kepler and Newton, orbits are ellipses in space, so your track and 3D projection are telling the truth — that if you stood back from the Earth and just watched the orbit, it would go in a rough ellipse (which would not be perfect because of the Moon and Sun and Jupiter and other gravitational effects, but would still be close).

But the Earth, as you watched, would be rotating on its axis — so each time the satellite came back to roughly the same point in its orbit, a different continent or city or ocean will have moved under that spot. So even though the satellite keeps reaching the ”same place” in space, it keeps finding a different part of the Earth beneath it.

My Skyfield astronomy package includes some routines that convert the SGP4 coordinates that are relative to the stars into ITRF coordinates that reflect what part of Earth the satellite is over:

https://github.com/brandon-rhodes/python-skyfield/blob/master/skyfield/sgp4lib.py

I have been wondering whether I should move these functions — now that they are written — over into the `sgp4` package itself, which is where they probably belong, since they are peculiar to, and work with, only SGP4 coordinates.

My reluctance has been because they (as currently written, at least) require NumPy arrays to do their matrix rotation, and I have not wanted the `sgp4` package to require outside libraries so that people can more easily use it in their projects. I wonder if I could somehow make the routines work without NumPy if someone did not have it installed?

**Comment by owntheweb** [comment]
Helllo @brandon-rhodes,
Thank you for the excellent feedback and explanation. Until I saw the link above and had further discussions with my co-workers who dabble more in this subject for education purposes, I  always thought orbits were not in a nice elipse like that (learned something new!). :D 

I started converting Javascript to Python from [this library](https://github.com/shashwatak/satellite-js) (which appears to have taken a lot of code directly from this project). 

![preview3-orbit-issue](https://cloud.githubusercontent.com/assets/611860/5780376/6249ac94-9d67-11e4-8df3-0edf5e36e8b6.png)

I've made some progress but due to ignorance in the area of what coordinate transforms to use when, I'm still off a bit (rendered image using an "eci_to_ecf" function that doesn't sound related to ITRF you suggested). I'll be sure to take a look at sgp4lib and the ITRF transform before proceeding with other endeavors.

Good thoughts on whether or not to merge sgp4lib with sgp4 or not. Perhaps one mild suggestion would be to mention sgp4lib on the sgp4 project page(s). That would prevent new dependencies for people already using sgp4 and also offer options that may be missed initially otherwise.

Thanks again!
